[AN] test: 설정, 프로필 화면 테스트

HamBeomJoon opened the pull request:
## 📌 변경 내용 & 이유
- SettingScreenTest
- ProfileScreenTest

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
#766 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Tests**
  * 프로필 화면 계측 테스트 추가(항목 표시, 기본 닉네임, 뒤로가기 동작 검증)
  * 설정 화면 테스트 확장(메뉴 표시, 로그인 상태별 UI, 로그인/로그아웃/회원탈퇴/내정보/뒤로가기 콜백 검증)

* **Style**
  * 프로필 및 하단 정보 간격 소폭 축소
  * 설정 메뉴 클릭 피드백을 비표시 리플 및 사용자 지정 상호작용으로 개선

* **New Features**
  * 크리스마스 테마 런처 아이콘 및 배경 리소스 추가 및 매니페스트 아이콘 교체
  * 앱 버전명/코드(1.4.1) 업데이트 및 관련 안내 문구 반영

* **Bug Fixes**
  * 뷰 바인딩 null 안전성 강화로 UI 시점 관련 문제 예방

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HamBeomJoon committed:
[AN] test: SettingScreenTest 로그인 상태별 UI 테스트 추가 및 보완 #766

*   [AN] test: 로그인하지 않은 경우 로그아웃 및 회원탈퇴 버튼 미노출 검증 추가
*   [AN] test: 로그인한 경우 로그아웃 및 회원탈퇴 버튼 노출 검증 테스트 추가

HamBeomJoon committed:
[AN] test: SettingScreen기본 항목 표시 테스트 #766

HamBeomJoon committed:
[AN] test: SettingScreen 동작 테스트 코드 추가 #766

*   [AN] test: 뒤로가기 버튼 클릭 시 콜백 호출 테스트 추가
*   [AN] test: 로그인, 로그아웃, 회원탈퇴 클릭 시 콜백 호출 테스트 추가
*   [AN] test: 내 정보(프로필) 클릭 시 콜백 호출 테스트 추가

HamBeomJoon committed:
[AN] test: ProfileScreen 테스트 코드 추가 #766

*   [AN] test: ProfileScreen UI 및 사용자 정보 표시 테스트(ProfileScreenTest.kt) 추가

HamBeomJoon committed:
[AN] refactor: settingItem 터치 리플효과 제거 #766

HamBeomJoon committed:
[AN] refactor: ProfileContent 마진 조절 #766

HamBeomJoon committed:
[AN] chore: ktlint 적용

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/setting/component/SettingContent.kt:
```diff
@@ -100,7 +102,11 @@ private fun SettingItem(
         modifier =
             modifier
                 .fillMaxWidth()
-                .clickable(onClick = onClick),
+                .clickable(
+                    onClick = onClick,
+                    indication = null,
+                    interactionSource = remember { MutableInteractionSource() },
+                ),
```
setting화면 항목 ripple제거했습니다.

rosemin928 commented on android/app/src/androidTest/java/com/onair/hearit/screen/ProfileScreenTest.kt:
```diff
@@ -0,0 +1,121 @@
+            )
+        }
+
+        composeTestRule.waitForIdle()
```
이건 왜 달아준건가여?? 진몰이라서 여쭤봅니다

HamBeomJoon replied:
```kotlin
// 1. assertion 없이 UI 상태만 확인할 때
composeTestRule.setContent { ... }
composeTestRule.waitForIdle()
val state = someState.value  // 직접 상태 읽기

// 2. 애니메이션 완료 대기
composeTestRule.mainClock.advanceTimeBy(500)
composeTestRule.waitForIdle()

// 3. LaunchedEffect 등 비동기 작업 완료 대기
composeTestRule.waitForIdle()
```

- 위와 같은 경우에 사용해야 한다고 하는데, 처음에 있었다가 테스트 코드들을 수정하고 하면서 지금은 필요없는 것 같네요!

HamBeomJoon replied:
```kotlin
// 이 메서드들이 내부적으로 idle 대기를 포함
onNodeWithText("...")      // 노드 찾을 때까지 대기
assertIsDisplayed()        // idle 대기 후 검증
assertDoesNotExist()       // idle 대기 후 검증
performClick()             // idle 대기 후 클릭
```

위 함수들은 waitForIdle이 필요없다고 하네요

https://github.com/woowacourse-teams/2025-hEARit/pull/769/commits/9b50569980d5988712084d08ed92feaebd320d24

rosemin928 commented on android/app/src/androidTest/java/com/onair/hearit/screen/ProfileScreenTest.kt:
```diff
@@ -0,0 +1,121 @@
+        composeTestRule.waitForIdle()
+
+        // then
+        composeTestRule
+            .onNodeWithText("내 정보")
+            .assertIsDisplayed()
```
저희는 현재 다국어를 지원하지 않고 있어서 아직 상관없을 것 같지만 만약 다국어를 지원하게 된다면, 지금처럼 문자열로 테스트를 진행하면 문제가 생길 수도 있을 것 같아요!
testTag를 달아주는 방법도 있었던 것 같은데, 또 프로덕션 코드에 테스트를 위한 코드가 섞이게 된다는 생각이 들면서 우테코 미션을 진행하면서도 고민을 한참 했던 기억이 있네요🥲

미플은 어떻게 생각하시나요?? 만약 히어릿이 다국어를 지원하게 된다면 어떤 식으로 테스트를 해야 좋을까요?

HamBeomJoon replied:
1. 리소스 ID로 문자열 가져오기 
2. contentDescription 사용

두 가지 방법이 있는 것 같은데, 텍스트는 이미 스크린 리더가 읽을 수 있어 contentDescription을 다는 것은 불필요하게 같은 내용을 두 번 읽는다고 하네요. 
`composeTestRule.activity.getString(R.string.setting_profile)` 와 같이 리소스 ID 방식을 채택했습니다. 
 - 닉네임, 앱 버전, 히어릿 이메일은 다국어랑 관련없음 + 고정값들이라 그냥 하드코딩으로 테스트했습니다.
- 아래쪽에 뒤로가기버튼은 Description으로 테스트를 진행했습니다.
```kotlin
 composeTestRule.onNodeWithContentDescription("뒤로가기").performClick()
```

https://github.com/woowacourse-teams/2025-hEARit/pull/769/commits/f5d9f3a26213d68ebd084dca00a1cda4a4c3a52a

rosemin928 commented on android/app/src/androidTest/java/com/onair/hearit/screen/SettingScreenTest.kt:
```diff
@@ -21,7 +24,6 @@ class SettingScreenTest {
     val composeTestRule = createComposeRule()
```
lifeCycleOwner, context 등을 사용하는 화면에 대한 테스트를 할 때는 `createComposeRule()`보다 `createAndroidComposeRule<ComponentActivity>()`를 사용하는 것이 더 실제 앱 실행 환경에 가깝게 테스트를 진행할 수 있다고 하는데 한 번 알아보셔도 좋을 것 같습니다!

HamBeomJoon replied:
- 오호.. 몰랐는데, Composable 내부에서 다음을 사용할 때 사용하면 좋다고 하는군요
<img width="732" height="222" alt="스크린샷 2025-12-22 15 23 50" src="https://github.com/user-attachments/assets/908c51b4-6a7a-44cb-80cd-f272e241a108" /> <br>

- 근데 찾아보니, createComposeRule함수가 내부적으로 createAndroidComposeRule<ComponentActivity>()를 구현하고 있네요  ㅋㅋㅋ
```kotlin
actual fun createComposeRule(): ComposeContentTestRule =
    createAndroidComposeRule<ComponentActivity>()

@ExperimentalTestApi
actual fun createComposeRule(effectContext: CoroutineContext): ComposeContentTestRule =
    createAndroidComposeRule<ComponentActivity>(effectContext)
```
그래서 테스트가 잘 통과되는 것 같네요. 

<img width="732" height="222" alt="스크린샷 2025-12-22 15 29 45" src="https://github.com/user-attachments/assets/511aa9d2-d3d8-46e7-b248-5645e9aaaefb" />

- 결국 테스트할때 Activity참조에 직접 접근하거나 커스텀 Activity를 사용하는 경우에 명시적으로 사용하면 될 것 같아요!

```kotlin
// Activity 인스턴스에 직접 접근해야 할 때
val context = composeTestRule.activity
val intent = composeTestRule.activity.intent
composeTestRule.activity.requestedOrientation = ...

// 커스텀 Activity로 테스트해야 할 때
createAndroidComposeRule<MyCustomActivity>()
```

HamBeomJoon replied:
위에 텍스트를 테스트하면서 `composeTestRule.activity.getString(R.string.setting_profile)` 가 필요해서 ProfileScreen은 createAndroidComposeRule로 만들었습니다.

rosemin928 reviewed (approved):
테스트 작성 수고하셨어요 미플! 저번에 제가 드린 코멘트 빠지지 않고 열심히 작성해주신 것 멋있네여 👍
개인적으로 궁금한 것과, 생각해보고 싶은 것들에 대해서 코멘트를 간단히 남겨두었으니 확인해주세요!

리뷰가 늦어져서 죄송합니다 🙇‍♀️

HamBeomJoon committed:
[AN] test: 불필요한 waitForIdle() 호출 제거 #766

Compose 테스트는 기본적으로 UI와 동기화되므로, 액션이나 단언을 호출하기 전에 UI가 유휴 상태가 될 때까지 자동으로 기다립니다. 이에 따라 중복으로 판단되는 `composeTestRule.waitForIdle()` 호출을 `SettingScreenTest`와 `ProfileScreenTest`에서 제거했습니다.

HamBeomJoon committed:
[AN] test: ProfileScreenTest에서 문자열 리소스 사용을 위해 Rule 변경 #766

*   [AN] test: createComposeRule을 createAndroidComposeRule로 변경하여 Context 접근 가능하도록 수정
*   [AN] test: 하드코딩된 "내 정보" 텍스트를 R.string.setting_profile 리소스 참조로 변경

HamBeomJoon committed:
[AN] style: 크리스마스 아이콘 변경 #766

HamBeomJoon committed:
[AN] fix: 빠른 탭 전환시 팅기는 버그 수정 #766

HamBeomJoon committed:
[AN] build: v1.4.1 버전 업그레이드 #766

HamBeomJoon committed:
Merge branch 'develop-an' into test/#766-setting-test
